lpc55_sign/src/bin/lpc55_sign.rs

```diff
@@ -182,7 +201,292 @@ fn main() -> Result<()> {
+    }
+
+    let expected_len =
+        header_offset + cert_header.header_length + cert_header.certificate_table_len + 32 * 4;
+    if cert_header.total_image_len != expected_len {
+        println!(
+            "❌ Invalid image length in cert header: expected {expected_len}, got {}",
+            cert_header.total_image_len
+        );
+    } else {
+        println!("✅ Verified certificate header length");
+    }
+
+    let mut start = (header_offset + cert_header.header_length) as usize;
+    let mut certs: Vec<x509_parser::certificate::X509Certificate> = vec![];
+    for i in 0..cert_header.certificate_count {
+        let x509_length = u32::from_le_bytes(image[start..start + 4].try_into().unwrap());
+        println!(
+            "Checking certificate [{}/{}]",
+            i + 1,
+            cert_header.certificate_count
+        );
+        if verbose {
+            println!("    certificate length: {x509_length}");
+        }
+        start += 4;
+        let cert = &image[start..start + x509_length as usize];
+
+        let (_, cert) = x509_parser::parse_x509_certificate(cert)?;
+        println!("    ✅ Successfully parsed certificate");
+
+        let prev_public_key = certs.last().map(|prev| prev.public_key());
+        let kind = if prev_public_key.is_some() {
+            "chained"
+        } else {
+            "self-signed"
+        };
+        match cert.verify_signature(prev_public_key) {
+            Ok(()) => println!("    ✅ Verified {kind} certificate signature"),
+            Err(e) => println!("    ❌ Failed to verify {kind} certificate signature: {e:?}"),
+        }
+
+        certs.push(cert);
+        start += x509_length as usize;
+
+        // TODO: verify that this certificate is signed by the previous one
+    }
+
+    let mut rkh_table = vec![];
+    let mut rkh_sha = sha2::Sha256::new();
+    for i in 0..4 {
+        let rot_hash = &image[start..start + 32];
+        if verbose {
+            print!("Root key hash {i}: ");
+            for r in rot_hash {
+                print!("{r:02x}");
+            }
+            println!();
+        }
+        rkh_sha.update(rot_hash);
+        rkh_table.push(rot_hash.to_owned());
+        start += 32;
+    }
+
+    if rkh_sha.finalize().as_slice() != cmpa.rotkh {
+        println!("❌ RKH in CMPA does not match Root Key hashes in image");
+    } else {
+        println!("✅ RKH in CMPA matches Root Key hashes in image");
+    }
+
+    let mut sha = sha2::Sha256::new();
+    let public_key = &certs[0].tbs_certificate.subject_pki.subject_public_key;
+    let public_key_rsa = rsa::RsaPublicKey::from_pkcs1_der(public_key.as_ref()).unwrap();
+    sha.update(public_key_rsa.n().to_bytes_be());
+    sha.update(public_key_rsa.e().to_bytes_be());
+    let out = sha.finalize().to_vec();
+    if !rkh_table.contains(&out) {
+        println!("❌ Certificate 0's public key is not in RKH table");
+    } else {
+        println!("✅ Certificate 0's public key is in RKH table");
+    }
+
```

Comment:
Need to check that RKH table slot matching Certificate 0's public key is enabled in CFPA

Reply:
Good idea, done